src/main/java/htsjdk/samtools/BAMShardIndexer.java

```diff
@@ -0,0 +1,190 @@
+
+          if (rec.getAlignmentStart() == SAMRecord.NO_ALIGNMENT_START) {
+              return; // do nothing for records without coordinates, but count them
+          }
+
+          // various checks
+          final int reference = rec.getReferenceIndex();
+          if (reference != currentReference) {
+              throw new SAMException("Unexpected reference " + reference +
+                      " when constructing index for " + currentReference + " for record " + rec);
+          }
+
+          binningIndexBuilder.processFeature(new BinningIndexBuilder.FeatureToBeIndexed() {
+              @Override
+              public int getStart() {
+                  return rec.getAlignmentStart();
+              }
+
+              @Override
+              public int getEnd() {
+                  return rec.getAlignmentEnd();
+              }
+
+              @Override
+              public Integer getIndexingBin() {
+                  final Integer binNumber = rec.getIndexingBin();
+                  return (binNumber == null ? rec.computeIndexingBin() : binNumber);
+
+              }
+
+              @Override
+              public Chunk getChunk() {
+                  final SAMFileSource source = rec.getFileSource();
+                  if (source == null) {
+                      throw new SAMException("No source (virtual file offsets); needed for indexing on BAM Record " + rec);
+                  }
+                  return ((BAMFileSpan) source.getFilePointer()).getSingleChunk();
+              }
+          });
+
+      }
+
+      /**
+       * Creates the BAMIndexContent for this reference.
+       * Requires all alignments of the reference have already been processed.
+       *
+       * @return Null if there are no features for this reference.
+       */
+      public BAMIndexContent processReference(final int reference) {
+
+          if (reference != currentReference) {
+              throw new SAMException("Unexpected reference " + reference + " when constructing index for " + currentReference);
+          }
+
+          final BinningIndexContent indexContent = binningIndexBuilder.generateIndexContent();
+          if (indexContent == null) return null;
+          return new BAMIndexContent(indexContent.getReferenceSequence(), indexContent.getBins(),
+                  indexStats, indexContent.getLinearIndex());
+
+      }
+
+      /**
+       * @return the count of records with no coordinate positions
+       */
+      public long getNoCoordinateRecordCount() {
+          return indexStats.getNoCoordinateRecordCount();
+      }
+
+      /**
+       * reinitialize all data structures when the reference changes
+       */
+      void startNewReference(int reference) {
+          currentReference  = reference;
+          // I'm not crazy about recycling this object, but that is the way it was originally written and
+          // it helps keep track of no-coordinate read count (which shouldn't be stored in this class anyway).
```

Comment:
So why not create a separate class to clean up the code?  You got my +1 vote :smile:  Less is always more in my book.